### PR TITLE
Fixes Issue #277. FCSData objects now able to be (de)pickled.

### PR DESCRIPTION
**Pull request by SynbioLucas**
Added code as described in Issue #277 to enable FCSData objects to be pickled and depickled. Wrote a test to verify that this is working properly and that depickled objects contain the correct attributes. All tests pass with no errors.

### CODE REVIEW

File: FlowCal/io.py
Comment (castillohair) [inline FlowCal/io.py]: Remove these. I would rather have this explained in a docstring in each function with more detail. See below.

File: FlowCal/io.py
Comment (castillohair) [inline FlowCal/io.py:1779]: We need a docstring here explaining what this function does and why it was replaced. It doesn't have to be too detailed, as this function will not be exposed to the user, but it should be enough to be useful for anyone in the future modifying this code. Please refer to https://github.com/numpy/numpy/blob/master/doc/example.py.

File: FlowCal/io.py
Comment (castillohair) [inline FlowCal/io.py:1844]: We need a docstring here explaining what this function does and why it was replaced. It doesn't have to be too detailed, as this function will not be exposed to the user, but it should be enough to be useful for anyone in the future modifying this code. Please refer to https://github.com/numpy/numpy/blob/master/doc/example.py.

File: test/test_io.py
Comment (castillohair) [inline test/test_io.py]: Why is this necessary? You should know in advance if these attributes are valid or not.

File: test/test_io.py
Comment (castillohair) [inline test/test_io.py:2716]: This section is checking that the same attributes exist, and that some subset of the hidden attributes have the same value. I would recommend that the contents of the array and the actual values of computed attributes are also checked. Even if you know that these worked before this PR, it would be a good idea for the tests to be as comprehensive as possible in order to detect future code changes that break proper pickling.

For the array values, you might want to use [`assert_array_equal`](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.testing.assert_array_equal.html) or [`assert_almost_equal`](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.testing.assert_almost_equal.html).

**Comment by castillohair** [comment]
Thanks for this PR. Indeed all old tests still pass and the Excel UI still works with the example file.

I'm requesting docstrings in the two functions that were added to `FCSData`, and a few additions to the test function.

**Review by castillohair — CHANGES_REQUESTED** [review]
Thanks for this PR. Indeed all old tests still pass and the Excel UI still works with the example file.

I'm requesting docstrings in the two functions that were added to `FCSData`, and a few additions to the test function. After that I think it should be ready to merge, unless @JS3xton has anything else to say.

File: test/test_io.py
Comment (SynbioLucas) [reply test/test_io.py:2716]: These are good ideas - thanks! I believe I've incorporated your suggestions, but please let me know if there are still cases/attributes/data that I'm forgetting!

File: test/test_io.py
Comment (SynbioLucas) [reply test/test_io.py]: Yep, you're right - this is now deleted.

**Comment by SynbioLucas** [comment]
Thanks for your great suggestions. It's obviously been a while since I've worked on this, but would love to still merge if it's possible. I believe I've implemented your requested improvements, but please feel free to let me know if you think of anything else!

Since it's been ages, I also merged in the current `taborlab/master` branch and confirmed that all 365 tests still pass (at least in Python 2.7...).

**Comment by castillohair** [comment]
Made some modifications so that tests pass on python3. They still work on python2.7. @JS3xton if you don't have any other comments, I'd like to merge this now.

**Review by castillohair — APPROVED** [review]

File: test/test_io.py
Comment (JS3xton) [inline test/test_io.py:13]: This `except` statement should be as precise as possible. On my computer, you should except a `ModuleNotFoundError`.

**Comment by SynbioLucas** [comment]
While I'm also a fan of specific exception cases, in this particular case,
I think it makes sense to fallback on standard `pickle` no matter what
error was encountered. That said, I don't know of any reason that `cPickle`
would fail to import since it's part of the standard library, but :shrug: I
wrote this way too long ago.

File: FlowCal/io.py
Comment (JS3xton) [reply FlowCal/io.py:1779]: Consider naming the returned parameters to hint at what they are, e.g.:

        Returns
        -------
        init : callable
            Callable used to recreate the inital instance
        init_args : tuple
            Arguments to be passed to the callable to reconstitute the
            FCSData object
        state : tuple
            Internal state data needed to fully reconstruct the instance. Data
            is passed to self.__setstate__() to reconstruct object during
            deserialization. The order of elements must match that expected
            by __setstate__.

Also, fix the typo with "inital"

File: FlowCal/io.py
Comment (JS3xton) [inline FlowCal/io.py]: These variable names were confusing to me. References to "state" appear most appropriate for the 3rd tuple element; calling the entire `__reduce__` tuple `pickled_state` therefore sounds imprecise to me.

The [Python 3 docs](https://docs.python.org/3.7/library/pickle.html#object.__reduce__) suggest the output of `__reduce__` is called a "reduce value":
> The __reduce__() method takes no argument and shall return either a string or preferably a tuple (the returned object is often referred to as the “reduce value”).

So consider changing `pickled_state` to something like `super_reduce_value`.

---

Also, the docs suggest that it's possible for `__reduce__` to return up to a 5-tuple, which this code would currently truncate. Can we guarantee that `np.ndarray` will never use those extra tuple elements? Or that it will always already have a "state" element in the 3rd position? I prefer a more precise implementation that is robust to `__reduce__` outputs with between 2 and 5 elements, as the docs allow for.

File: FlowCal/io.py
Comment (JS3xton) [inline FlowCal/io.py]: The [docs](https://docs.python.org/3.7/library/pickle.html#object.__reduce__) don't specify how the "state" information is  stored (i.e. in general, we don't know what kind of object `pickled_state[2]` is). For now, the `+` operator appears to append our state information to the parent's state information, but I think it would be more robust to layer our own state information around the parent state information without making any assumptions about the parent state information. E.g. `new_state = (pickled_state[2],fcsdata_attrs)`. Our `__setstate__` function would therefore set our attributes from `new_state[1]` and then pass `new_state[0]` to the `super.__setstate__`.

If we're separating our own state information, you might also consider using a `namedtuple` or `dict` to avoid relying on positional indices to get the attributes unmixed correctly.

File: FlowCal/io.py
Comment (JS3xton) [inline FlowCal/io.py]: As mentioned in `__reduce__`, consider a more robust implementation that packages our state information separate from the parent state information and uses better data structures to avoid relying on position/index to unpack our attributes.

Comment (JS3xton) [reply FlowCal/io.py]: Also, does it matter who does `__setstate__` first? Practically, I think it doesn't matter for this case (?), but philosophically, I kind of feel like the parent class should go first, but I'm not sure.

File: FlowCal/io.py
Comment (castillohair) [reply FlowCal/io.py:1779]: [Name of return values is optional](https://numpydoc.readthedocs.io/en/latest/format.html), as they're not exposed to the method caller anyway.

**Comment by castillohair** [comment]
@JS3xton, do you wanna make the changes you proposed?

**Comment by JS3xton** [comment]
Alright, I took a stab. It currently passes all unit tests on my machine.

**Pull request merged**
